### PR TITLE
Beam Search optimization trainer for PNP

### PR DESCRIPTION
**Pull request by jayantk**
New training algorithm that optimizes a globally-normalized model by training the weights to ensure that  correct program executions don't fall off the beam.

Also some random refactoring and code cleanup.

### CODE REVIEW

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        queue.correctQueue.getItems.slice(
+            0, queue.correctQueue.size).maxBy(x => x.logProb)
+      } else {
+        null
+      }
+      log.stopTimer("bso/beam_search/get_executions")
+
+      var nextBeamSize = -1
+      if (numIters != 0 && bestCorrectEx != null && worstIncorrectEx != null &&
+          worstIncorrectEx.logProb + 1 > bestCorrectEx.logProb) {
```
Comment (OyvindTafjord) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: Maybe make a named constant of the "magic" 1 here and wherever else it's used (like line 104). Would one ever want to try another constant here?

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        // Margin violation
+        // println("m: " + numIters + " " + worstIncorrectEx.logProb + " " + bestCorrectEx.logProb)
+
+        // Add to the loss.
+        log.startTimer("bso/beam_search/margin_violation")
+        
+        val beamScoreExpr = worstIncorrectEx.env.getScore(false)
+        val correctScoreExpr = bestCorrectEx.env.getScore(false)
+
+        // XXX: allow different costs.
```
Comment (OyvindTafjord) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: Ah, regarding previous comment, looks like this is a todo already

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+      
+      trainer.update_epoch()
+
+      log.logStatistic(i, "loss", loss)
+      log.notifyIterationEnd(i)
+    }
+  }
+  
+  private def doExampleUpdate[A](example: PpExample[A]): Double = {
+    val cg = new ComputationGraph
```
Comment (joelgrus) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: prefer `ComputationGraph.getNew`

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    val finished = new BsoPpQueue[A](beamSize, stateCost, graph, log)
+
+    val startEnv = env.setLog(log)
+    queue.offer(example.unconditional, env, 0.0, null, null, env)
+
+    log.startTimer("bso/beam_search")
+    val beam = new Array[SearchState[A]](beamSize)
+    var numIters = 0
+    val losses = ListBuffer[Expression]()
+    while (queue.queue.size > 0 &&
```
Comment (joelgrus) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: `queue.queue.size` feels awkward to me; should maybe `BsoPpQueue` have its own `size` method rather than having to know about its internals?

Comment (joelgrus) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: eh, I guess if it's never used outside this file it doesn't matter

File: src/main/scala/org/allenai/pnp/GlobalLoglikelihoodTrainer.scala [hunk 5]
```diff
@@ -32,47 +35,41 @@ class GlobalLoglikelihoodTrainer(val epochs: Int, val beamSize: Int,
...
-        trainer.update(1.0f)
         cg.delete()
       }
+      logFn.logStatistic(i, "search errors", searchErrors) 
       // println(i + "  loss: " + loss)
       trainer.update_epoch()
     }
   }
   
   private def marginalsToLogProbExpression[A](marginals: PpBeamMarginals[A]): Expression = {
```
Comment (joelgrus) [inline src/main/scala/org/allenai/pnp/GlobalLoglikelihoodTrainer.scala]: `Option[Expression]` would be more scala-y

Comment (OyvindTafjord) [inline src/main/scala/org/allenai/pnp/GlobalLoglikelihoodTrainer.scala]: Right, ideally there should be no `null`s anywhere in a Scala program.

**Comment by joelgrus** [comment]
I think changing to `ComputationGraph.getNew` is important, the other comments are your call

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala (hunk 1 above)
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: Yeah ideally client code will be able to select the value instead of 1, i.e., the margin. I have to figure out the right interface for implementing cost functions, then we can fix this.

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala (hunk 4 above)
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: The queue has several different queues in it, each of which has its own size, so I think this way makes sense. We're never going to use the class outside this file.

File: src/main/scala/org/allenai/pnp/GlobalLoglikelihoodTrainer.scala (hunk 5 above)
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/GlobalLoglikelihoodTrainer.scala]: fixed

**Comment by jayantk** [comment]
I changed everything to computationGraph.getNew here and in all other code in pnp.
